### PR TITLE
ARROW-9005: [Rust] [Datafusion] support sort expression

### PR DESCRIPTION
**Pull request by houqp**

### CODE REVIEW

**Comment by houqp** [comment]
@nevi-me do you prefer to move concat kernel logic into append_data in this set of change or as a follow up after https://github.com/apache/arrow/pull/7306 get merged?

**Comment by nevi-me** [comment]
> @nevi-me do you prefer to move concat kernel logic into append_data in this set of change or as a follow up after #7306 gets merged?

I've opened https://issues.apache.org/jira/browse/ARROW-9007, so we can address `append_data` as part of that

**Comment by andygrove** [comment]
@houqp There is a 0.2 branch in sqlparser-rs where you can apply your changes and I can make a new 0.2.x release

**Comment by andygrove** [comment]
This looks great @houqp. I will review in detail in the next day or two.

File: rust/arrow/src/compute/kernels/concat.rs
Comment (nevi-me) [inline rust/arrow/src/compute/kernels/concat.rs:56]: `... with different data types`

File: rust/arrow/src/compute/kernels/concat.rs
Comment (nevi-me) [inline rust/arrow/src/compute/kernels/concat.rs:137]: do you mind changing these to `&[ArrayDataRef]`? There was a PR a few weeks ago to change borrowed vecs to slices, I think it's one of the clippy lints

File: rust/arrow/src/compute/kernels/concat.rs
Comment (nevi-me) [inline rust/arrow/src/compute/kernels/concat.rs:189]: Would this function handle array slices correctly? May you please test this, either in a new test case or by slicing the second string array in `test_concat_string_arrays`?

File: rust/arrow/src/compute/kernels/sort.rs
Comment (nevi-me) [inline rust/arrow/src/compute/kernels/sort.rs:125]: Why the change here?

**Review by nevi-me — CHANGES_REQUESTED** [review]
Minor suggestions from my side, I'll let @andygrove review the sort changes in DataFusion comprehensively, otherwise everything looks fine.

`concat` support is exciting, thanks @houqp

File: rust/arrow/src/compute/kernels/concat.rs
Comment (houqp) [reply rust/arrow/src/compute/kernels/concat.rs:137]: good catch. I have updated it.

File: rust/arrow/src/compute/kernels/sort.rs
Comment (houqp) [reply rust/arrow/src/compute/kernels/sort.rs:125]: this is to match spark's default behavior, i have added a comment.

File: rust/arrow/src/compute/kernels/concat.rs
Comment (houqp) [reply rust/arrow/src/compute/kernels/concat.rs]: By handling array slices did you mean StringArray created from str array slice?

**Comment by houqp** [comment]
@andygrove since you mentioned 0.2 branch, do you expect datafusion to migrate to latest sqlparser-rs in the near future or are we going to stay with 0.2 for a long time? I already started a local branch to for the migration, but if we do want to stay on 0.2 release, then I should probably stop working on that ;)

File: rust/arrow/src/compute/kernels/sort.rs
Comment (nevi-me) [reply rust/arrow/src/compute/kernels/sort.rs:126]: Okay, that's fine, we could set the default in DataFusion instead, but the way you've done it is fine. I would have taken that as a lazy approach to avoid changing all the test cases :)

File: rust/arrow/src/compute/kernels/concat.rs
Comment (nevi-me) [reply rust/arrow/src/compute/kernels/concat.rs]: Sorry, I meant slicing the second string array, something like below, which slices out "1" and "6" from the second array:

```rust
    fn test_concat_string_arrays() -> Result<()> {
        let arr = concat(&vec![
            Arc::new(
                StringArray::try_from(vec![Some("hello"), Some("world")])
                    .expect("Unable to create string array"),
            ) as ArrayRef,
            StringArray::from(vec!["1", "2", "3", "4", "6"]).slice(1, 3),
            Arc::new(
                StringArray::try_from(vec![Some("foo"), Some("bar"), None, Some("baz")])
                    .expect("Unable to create string array"),
            ) as ArrayRef,
        ])?;

        let expected_output = Arc::new(
            StringArray::try_from(vec![
                Some("hello"),
                Some("world"),
                // Some("1"),
                Some("2"),
                Some("3"),
                Some("4"),
                // Some("6"),
                Some("foo"),
                Some("bar"),
                None,
                Some("baz"),
            ])
            .expect("Unable to create string array"),
        ) as ArrayRef;

        assert!(
            arr.equals(&(*expected_output)),
            "expect {:#?} to be: {:#?}",
            arr,
            &expected_output
        );

        Ok(())
    }
```

The same might need to be verified with numeric arrays because we write the whole data in `write_slice`, so if there's an offset, we might get unexpected outputs.

I'll check your branch out locally and check this.

Comment (nevi-me) [reply rust/arrow/src/compute/kernels/concat.rs]: I've checked, we get this failure below:

The same test failure occurs if you slice a numeric array.

```
thread 'compute::kernels::concat::tests::test_concat_string_arrays' panicked at 'expect StringArray
[
  "hello",
  "world",
  "1",
  "2",
  "3",
  "46f",
  "oob",
  null,
  "arb",
] to be: StringArray
[
  "hello",
  "world",
  "2",
  "3",
  "4",
  "foo",
  "bar",
  null,
  "baz",
]', arrow/src/compute/kernels/concat.rs:296:9
```

Comment (nevi-me) [reply rust/arrow/src/compute/kernels/concat.rs]: I'm nearly done with `append_data()`, which handles offsets. Is it worthwhile to wait for it? I can put up a draft PR tonight

Comment (nevi-me) [reply rust/arrow/src/compute/kernels/concat.rs]: @houqp here's a WIP of the changes: https://github.com/apache/arrow/compare/master...nevi-me:ARROW-9007

I'm still testing the string and nested array types, but the primitive types behave fine with offsets

Comment (houqp) [reply rust/arrow/src/compute/kernels/concat.rs]: Yeah, let's wait for that. I was going to refactor this code by implementing append_data. Good to avoid duplicate efforts :D

**Comment by andygrove** [comment]
@houqp Please take a look at https://issues.apache.org/jira/browse/ARROW-8824 regarding the SQL parser conversation

File: rust/arrow/src/compute/kernels/concat.rs
Comment (nevi-me) [reply rust/arrow/src/compute/kernels/concat.rs]: @houqp I need concat support, so I'm going to temporarily merge your code into a fork of mine. Hopefully we can merge this, #7306  and #7365 in the coming days :)

**Comment by houqp** [comment]
@andygrove sent PR to sqlparser: https://github.com/andygrove/sqlparser-rs/pull/185.

I recommend we focus on reviewing and getting https://github.com/apache/arrow/pull/7365 merged first. Then I can refactor this PR to use append_data methods.

File: rust/arrow/src/compute/kernels/concat.rs
Comment (houqp) [reply rust/arrow/src/compute/kernels/concat.rs]: great work on #7365, looking forward to refactor/simplify this PR after it's merged!

**Comment by andygrove** [comment]
@houqp I released sqlparser 0.2.6

**Comment by houqp** [comment]
alright, i have added nulls ordering support through sqlparser 0.2.6. only thing left is waiting for merge of #7365.

File: rust/arrow/src/array/builder.rs
Comment (houqp) [inline rust/arrow/src/array/builder.rs:1126]: move append_data out of the for loop so memory allocation count for buffers is constant.

**Comment by houqp** [comment]
@nevi-me @andygrove rebased and ready for review

File: rust/arrow/src/array/array.rs
Comment (nevi-me) [inline rust/arrow/src/array/array.rs:1312]: For consistency, may you please also add `builder` to `BinaaryArray`. They both behave the same way

File: rust/arrow/src/compute/kernels/concat.rs
Comment (nevi-me) [inline rust/arrow/src/compute/kernels/concat.rs:66]: If you change this to check that `!array_list.is_empty()` at the start of the function, you could then get the datatype from the first element in `array_list`, then avoid validating data types as `append_data` will do that for us

File: rust/arrow/src/compute/kernels/concat.rs
Comment (nevi-me) [inline rust/arrow/src/compute/kernels/concat.rs:69]: typo, `cocat` -> `concat`

File: rust/arrow/src/compute/kernels/concat.rs
Comment (nevi-me) [inline rust/arrow/src/compute/kernels/concat.rs:138]: It's better to return `Err(ArrowError)` instead of panicking here

File: rust/datafusion/src/execution/physical_plan/sort.rs
Comment (nevi-me) [inline rust/datafusion/src/execution/physical_plan/sort.rs:147]: More a question for @andygrove; in the case of a sort operation consolidating all partitions into a single larger partition, would we need to repartition the record batches?

**Review by nevi-me — COMMENTED** [review]
Thanks, left some comments

**Review by nevi-me — APPROVED** [review]
Thanks @houqp

File: rust/datafusion/src/execution/physical_plan/sort.rs
Comment (houqp) [reply rust/datafusion/src/execution/physical_plan/sort.rs:147]: I have the same question while developing this, curious what @andygrove thoughts on this as well. I left it as one single batch for performance reason (to avoid having to run the split code).

**Comment by nevi-me** [comment]
We can address any queries in follow up Jiras

**Pull request closed without merging**
